packages/eslint-plugin/src/rules/consistent-return.ts

```diff
@@ -0,0 +1,95 @@
+import { type TSESTree } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import type {
+  InferMessageIdsTypeFromRule,
+  InferOptionsTypeFromRule,
+} from '../util';
+import { createRule, getParserServices, isTypeFlagSet } from '../util';
+import { getESLintCoreRule } from '../util/getESLintCoreRule';
+
+const baseRule = getESLintCoreRule('consistent-return');
+
+type Options = InferOptionsTypeFromRule<typeof baseRule>;
+type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
+
+type FunctionNode =
+  | TSESTree.FunctionDeclaration
+  | TSESTree.FunctionExpression
+  | TSESTree.ArrowFunctionExpression;
+
+export default createRule<Options, MessageIds>({
+  name: 'consistent-return',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Require `return` statements to either always or never specify values',
+      extendsBaseRule: true,
+      requiresTypeChecking: true,
+    },
+    hasSuggestions: baseRule.meta.hasSuggestions,
+    schema: baseRule.meta.schema,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+    const rules = baseRule.create(context);
+
+    let functionNode: FunctionNode | null;
+
+    function setFunctionNode(node: FunctionNode): void {
+      functionNode = node;
+    }
+
+    function isReturnPromiseVoid(
+      node: FunctionNode,
+      signature: ts.Signature,
+    ): boolean {
+      const tsNode = services.esTreeNodeToTSNodeMap.get(node);
+      const returnType = signature.getReturnType();
+      if (
+        tsutils.isThenableType(checker, tsNode, returnType) &&
+        tsutils.isTypeReference(returnType)
+      ) {
+        const typeArgs = returnType.typeArguments;
+        const hasVoid = !!typeArgs?.some(typeArg =>
+          isTypeFlagSet(typeArg, ts.TypeFlags.Void),
+        );
+        return hasVoid;
+      }
+      return false;
+    }
+
+    function isReturnVoidOrThenableVoid(node: FunctionNode): boolean {
+      const functionType = services.getTypeAtLocation(node);
+      const callSignatures = functionType.getCallSignatures();
```

Comment:
[perf] It fetches type information on every return statement, even within the same function, so 100 return statements === 100 `getTypeAtLocation` calls. Maybe we can somehow lazily get this type once and then reuse it from cache? What do you think?